drivers/*: unify drivers params definition scheme

aabadie opened the pull request:
This PR is an attempt to address what was suggested in #7519 and especially the first step of this [comment](https://github.com/RIOT-OS/RIOT/issues/7519#issuecomment-325584959).

~~I started to update some of the drivers also available in `auto_init/saul` but before going further, I'd like to know if I'm on the right way.~~

aabadie commented:
Finally, I updated all drivers and the ones with a saul adaption. In the mean time, I also removed some `*_params.h` files for some boards because I think there are useless since some values are already defined in the `board.h` and then are taken in to account by `*_params.h` in the drivers directory.

haukepetersen commented:
Nice effort! I think most of the changes are very valid and its nice to see the style of parameter definition to be cleaned up.

One think I don't like, that this PR is still using (slightly) different styles in the same context. For example looking at the `fox` board: the parameters for the radio are defined using some kind of static struct initializer style
```c
#define AT86RF2XX_PARAMS    { .spi = SPI_DEV(0),                \
                              ...
                              .reset_pin = GPIO_PIN(PORT_C, 1)}
```
while the on-board sensors are configured using singular defines
```c
 #define L3G4200D_I2C        I2C_0
 #define L3G4200D_ADDR       0x68
 #define L3G4200D_DRDY       GPIO_PIN(PORT_B,8)
 #define L3G4200D_INT        GPIO_PIN(PORT_B,11)
```

So I would prefer to keep to **one** style. IMHO the singular defines are the nicest, as this allows us to use default values which we don't need to re-define. So taking the `l3g4200d` as example, The `fox` configuration for this sensor would look like:
```c
/**
 * @name    Define the interface for the L3G4200D gyroscope
 * @{
 */
#define L3G4200D_PARAM_DRDY       GPIO_PIN(PORT_B,8)
#define L3G4200D_PARAM_INT        GPIO_PIN(PORT_B,11)
/** @} */
```
The I2C dev and addr values then simply fallback to their default values `I2C_DEV(0)` and `0x68`. NOTE: the `l3g4200_params.h` file needs also to be adapted as it does not define default values currently...

The same would be true for the `at86rf` radio, whose configuration would look like this:
```c
/**
 * @name    Define the interface to the AT86RF231 radio
 */
#define AT86RF2XX_PARAM_CS          (GPIO_PIN(PORT_A, 1))
#define AT86RF2XX_PARAM_INT         (GPIO_PIN(PORT_C, 2))
#define AT86RF2XX_PARAM_SLEEP       (GPIO_PIN(PORT_A, 0))
#define AT86RF2XX_PARAM_RESET       (GPIO_PIN(PORT_C, 1))
```
Here again: SPI dev and clock speed use the default values...

aabadie commented:
Thanks for having a look @haukepetersen.

> the singular defines are the nicest, as this allows us to use default values which we don't need to re-define

I totally agree with this and will apply the required change.

aabadie commented:
After looking at the board params (iotlab-m3, fox, etc), the actual status of this is leaving them broken. I'm working on fixing them...

aabadie commented:
@haukepetersen, I'm nearly done with this one (at least the `board.h` are ready). Can you have a look again ?

aabadie commented:
Ok, after some fight, CI is finally green here... but the diff has increased significantly.

@gebart @haukepetersen, can you check that all mulle auto configured devices are still working ?
Is there someone with `pba-d-01-kw2x` and `fox` boards that can try ?

It's ok on iotlab-m3.

aabadie commented:
Maybe @smlng is also interested in this one ?

aabadie commented:
@haukepetersen, do you think you'll have time soon to re-review this one ? I think it's ready now .

aabadie commented on drivers/bmx280/include/bmx280_params.h:
```diff
@@ -41,50 +41,46 @@ extern "C" {
-        .t_sb = BMX280_SB_0_5,             \
-        .filter = BMX280_FILTER_OFF,       \
+#ifndef BMX280_PARAMS
+#define BMX280_PARAMS                       \
```
Yes could be. I'll apply this change.

aabadie commented:
> I think it does however make sense to add the preferred style for this, and also for defining struct data, to the coding conventions, so we have something to point people to and so that we all can converge on the same style...

Agreed, maybe I can remove the parenthesis, since 2 (important) people are complaining about them.

haukepetersen commented on drivers/at86rf2xx/include/at86rf2xx_params.h:
```diff
@@ -49,24 +49,22 @@ extern "C" {
 
-#define AT86RF2XX_PARAMS_DEFAULT    {.spi = AT86RF2XX_PARAM_SPI, \
+#ifndef AT86RF2XX_PARAMS
+#define AT86RF2XX_PARAMS            {.spi = AT86RF2XX_PARAM_SPI, \
```
space missing

haukepetersen commented on drivers/bh1750fvi/include/bh1750fvi_params.h:
```diff
@@ -40,21 +40,19 @@ extern "C" {
 
-#define BH1750FVI_PARAMS_DEFAULT    {.i2c = BH1750FVI_PARAM_I2C, \
+#ifndef BH1750FVI_PARAMS
+#define BH1750FVI_PARAMS            {.i2c = BH1750FVI_PARAM_I2C, \
```
space missing

haukepetersen commented on drivers/grove_ledbar/include/grove_ledbar_params.h:
```diff
@@ -54,33 +54,34 @@ extern "C"
  */
-#define GROVE_LEDBAR_PARAMS {       \
+#ifndef GROVE_LEDBAR_PARAMS
+#define GROVE_LEDBAR_PARAMS     {   \
```
different style here.

haukepetersen commented on drivers/io1_xplained/include/io1_xplained_params.h:
```diff
@@ -38,32 +38,32 @@ extern "C" {
+#define IO1_XPLAINED_PARAMS           { .addr = IO1_XPLAINED_PARAM_ADDR }
+#endif
+#ifndef IO1_XPLAINED_SAUL_INFO
+#define IO1_XPLAINED_SAUL_INFO        {        \
```
different style

haukepetersen commented on drivers/kw2xrf/include/kw2xrf_params.h:
```diff
@@ -48,22 +48,20 @@ extern "C" {
 
-#define KW2XRF_PARAMS_DEFAULT    {.spi = KW2XRF_PARAM_SPI, \
+#ifndef KW2XRF_PARAMS
+#define KW2XRF_PARAMS            {.spi = KW2XRF_PARAM_SPI,         \
```
space missing

haukepetersen commented on drivers/lpd8808/include/lpd8808_params.h:
```diff
@@ -39,21 +39,19 @@ extern "C" {
 
-#define LPD8808_PARAMS_DEFAULT      {.led_cnt = LPD8808_PARAM_LED_CNT, \
+#ifndef LPD8808_PARAMS
+#define LPD8808_PARAMS              {.led_cnt = LPD8808_PARAM_LED_CNT, \
```
space missing

haukepetersen commented on drivers/lsm303dlhc/include/lsm303dlhc_params.h:
```diff
@@ -47,7 +47,7 @@ extern "C" {
 #endif
 #ifndef LSM303DLHC_PARAM_MAG_ADDR
-#define LSM303DLHC_PARAM_MAG_ADDR       (0x1e)
+#define LSM303DLHC_PARAM_MAG_ADDR       (0x1E)
```
just noticed: I think in the majority of the RIOT code base we use small letters for hexadecimal numbers, aren't we?!

But doing `git grep 0x`, it seems that both are used. This might also be something to put in the coding conventions I guess, but we would need to vote about something.

Anyway, no blocker for this PR. I think I will simply open an issue for voting on a style so we can converge in the future...

haukepetersen replied:
done: #8151

aabadie replied:
I would personally go for the capital letters but I agree this is not consistent in the code base.

haukepetersen commented on sys/auto_init/saul/auto_init_dht.c:
```diff
@@ -29,12 +29,22 @@
  * @brief   Define the number of configured sensors
  */
-#define DHT_NUMOF   (sizeof(dht_params) / sizeof(dht_params[0]))
+#define DHT_NUM   (sizeof(dht_params) / sizeof(dht_params[0]))
```
indention now off by 2

haukepetersen commented on sys/auto_init/saul/auto_init_hdc1000.c:
```diff
@@ -41,6 +41,11 @@ static hdc1000_t hdc1000_devs[HDC1000_NUM];
+/**
+ * @brief   Define the number of saul info
+ */
+#define HDC1000_INFO_NUM    (sizeof(hdc1000_saul_info)/sizeof(hdc1000_saul_info[0]))
```
spaces around `/` missing

haukepetersen commented on sys/auto_init/saul/auto_init_isl29020.c:
```diff
@@ -41,6 +41,11 @@ static isl29020_t isl29020_devs[ISL29020_NUM];
+/**
+ * @brief   Define the number of saul info
+ */
+#define ISL29020_INFO_NUM    (sizeof(isl29020_saul_info)/sizeof(isl29020_saul_info[0]))
```
spaces missing

haukepetersen commented on sys/auto_init/saul/auto_init_jc42.c:
```diff
@@ -28,17 +28,22 @@
  * @brief   Define the number of configured sensors
  */
-#define JC42_NUMOF    (sizeof(jc42_params) / sizeof(jc42_params[0]))
+#define JC42_NUM    (sizeof(jc42_params) / sizeof(jc42_params[0]))
```
indention

haukepetersen commented on sys/auto_init/saul/auto_init_jc42.c:
```diff
@@ -28,17 +28,22 @@
+/**
+ * @brief   Define the number of saul info
+ */
+#define JC42_INFO_NUM    (sizeof(jc42_saul_info)/sizeof(jc42_saul_info[0]))
```
spaces

haukepetersen commented on sys/auto_init/saul/auto_init_l3g4200d.c:
```diff
@@ -41,6 +41,11 @@ static l3g4200d_t l3g4200d_devs[L3G4200D_NUM];
+/**
+ * @brief   Define the number of saul info
+ */
+#define L3G4200D_INFO_NUM    (sizeof(l3g4200d_saul_info)/sizeof(l3g4200d_saul_info[0]))
```
spaces (also check files below...)

aabadie replied:
> spaces (also check files below...)

ok, I fixed all of them...

haukepetersen commented on sys/auto_init/saul/auto_init_tsl2561.c:
```diff
@@ -28,17 +28,22 @@
  * @brief   Define the number of configured sensors
  */
-#define TSL2561_NUMOF    (sizeof(tsl2561_params) / sizeof(tsl2561_params[0]))
+#define TSL2561_NUM    (sizeof(tsl2561_params) / sizeof(tsl2561_params[0]))
```
indention

haukepetersen commented:
sorry for still poring in change requests, but a PR this size you can look at a 100 times and still find something that you overlooked before...

aabadie commented:
That's fine, at least I have a review and it goes in the right direction :)

aabadie commented:
new bunch of comments addressed @haukepetersen

aabadie commented:
@haukepetersen rebased and squashed. I hope someone will find time to review this one again.

aabadie commented:
Fixed the conflicts and CI issues

aabadie commented:
@haukepetersen, I just rebased this one to fix conflicts.

aabadie commented:
> I agree with @smlng, that the extra parenthesis around define values as GPIO_PIN(x,y) are not very nice to look at and don't really add anything from a functional perspective

I just reverted this change.

aabadie commented:
@haukepetersen @smlng do you have more comments on this one ?

smlng commented:
@haukepetersen please check if you comments are addressed, so this can go into the release ,THX.

aabadie commented on boards/iotlab-a8-m3/include/board.h:
```diff
@@ -34,22 +34,16 @@ extern "C" {
-#define L3G4200D_ADDR       0x68
-#define L3G4200D_DRDY       GPIO_PIN(PORT_C,9)
-#define L3G4200D_INT        GPIO_PIN(PORT_C,6)
+#define L3G4200D_PARAM_INT2       GPIO_PIN(PORT_C, 9)
```
about LSM303DLHC_PARAM_ACC_PIN, I think there was a mismatch. Otherwise I tried to apply the same rule `<driver_name>_PARAM_<param>`

aabadie replied:
needs to be double checked though

kYc0o replied:
> needs to be double checked though

Was it?

aabadie replied:
[yes](https://github.com/RIOT-OS/RIOT/pull/7937#discussion_r161790688)

aabadie commented on boards/iotlab-a8-m3/include/board.h:
```diff
@@ -34,22 +34,16 @@ extern "C" {
-#define LSM303DLHC_ACC_ADDR (0x19)
-#define LSM303DLHC_MAG_ADDR (0x1e)
-#define LSM303DLHC_INT1     GPIO_PIN(PORT_B,12)
-#define LSM303DLHC_INT2     GPIO_PIN(PORT_B,2)
```
Yes, I think the default values are the same, or something like this, to be verified though

aabadie replied:
I just verified and yes it was unsused. That's why I removed it

aabadie commented:
ping @haukepetersen

smlng commented on boards/common/iotlab/include/board_common.h:
```diff
@@ -49,46 +49,46 @@ extern "C" {
-#define XTIMER_WIDTH        (16U)
-#define XTIMER_OVERHEAD     (6U)
-#define XTIMER_SHOOT_EARLY  (3U)
+#define XTIMER_WIDTH              (16U)
```
this is inconsistent with other changes here, and also with existing configs in RIOT  which are defined without `U`, hence signed.

smlng replied:
below, too

aabadie replied:
Fixed

aabadie commented:
@smlng I addressed your comment about signed values and fixed merge conflict in the meantime.

smlng commented:
@haukepetersen are we good here?

aabadie commented:
@haukepetersen, if you have no more comment, please stop blocking this PR. Thanks!

aabadie commented:
@haukepetersen any chance to get an ACK (or a review) here ?

aabadie commented:
@haukepetersen, just a reminder for this one. If you think it's too large to review, I can split in smaller PRs, but there will be quite a few if I do that...

aabadie commented:
Let's keep pushing @haukepetersen ;)

aabadie commented:
Since @haukepetersen is scanning his backlog this morning, let's try another time here. After that I think I'll consider dismissing his review, since there's already an ACK from @smlng

haukepetersen commented:
on it :-)

haukepetersen commented on boards/common/iotlab/include/board_common.h:
```diff
@@ -49,46 +49,46 @@ extern "C" {
-#define LED2_PIN            GPIO_PIN(PORT_C, 10)
+#define LED0_PIN                  GPIO_PIN(PORT_D, 2)
+#define LED1_PIN                  GPIO_PIN(PORT_B, 5)
+#define LED2_PIN                  GPIO_PIN(PORT_C, 10)
```
here and in (probably) many other places of this PR: the indention is not aligned to 4-spaces anymore... It was before, but now the indention changed by 6 space chars

haukepetersen commented on boards/mulle/include/board.h:
```diff
@@ -108,54 +108,53 @@ void board_init(void);
+#define LIS3DH_PARAM_INT1               GPIO_PIN(PORT_C, 18)
+#define LIS3DH_PARAM_INT2               GPIO_PIN(PORT_C, 17)
+#define LIS3DH_PARAM_CS                 SPI_HWCS(0)
+#define LIS3DH_PARAM_CLK                (SPI_CLK_5MHZ)
```
just noticed: in all the other places we removed the parenthesis around non-numeric values (e.g. the three lines above). So shouldn't this be the case here, too?

aabadie replied:
I don't think so

aabadie replied:
Regarding those parenthesis, this PR just applies the same 'rule' that is already used with other drivers (at86rf2xx and w5100 at least). I prefer that we keep them with simple defines like this but don't when using a macro function (such as GPIO_PIN, etc). What do you think @haukepetersen ?

haukepetersen commented on boards/common/iotlab/include/board_common.h:
```diff
@@ -49,46 +49,46 @@ extern "C" {
-#define XTIMER_WIDTH        (16U)
-#define XTIMER_OVERHEAD     (6U)
-#define XTIMER_SHOOT_EARLY  (3U)
+#define XTIMER_WIDTH              (16)
```
changing `16U` to `16` should not be part of this PR...

aabadie replied:
It was requested by @smlng [some times ago](https://github.com/RIOT-OS/RIOT/pull/7937#discussion_r162161669), I made that consistent across other board configurations.

aabadie commented:
> you broke the 4-space indention in most files

Is this a new rule ?

haukepetersen commented:
> Is this a new rule ?

I always thought its an old rule. Align using tabs but use 4 spaces instead of a tab character, is it not?

haukepetersen commented:
Just checked the [coding conventions](https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions):

> Indentations are four spaces (i.e., NO tab characters).

I wonder if it would simplify this specific issue if we merge this PR with broken intention and you fix the indention in a follow-up? That would be fine with me.

aabadie commented:
> Indentations are four spaces (i.e., NO tab characters).

I understand this as spaces from the beginning of the line not between code and doxygen comments.
It's also unclear if aligning a splitted line:
```
void example_function(int arg1, int arg2,
                      int arg3)  /* indent is multiple of four */
{
     ...
}

or 

void example_function(int arg1, int arg2,
                    int arg3) /* indent is multiple of four */
{
     ...
}
```
?

aabadie commented:
> I wonder if it would simplify this specific issue if we merge this PR with broken intention and you fix the indention in a follow-up?

I won't do that since there's no broken indentation introduced by this PR. My editor is configured in a way that converts tabs in 4 spaces automatically. The 6 **spaces** (NO tab) added were just there to align the doxygen comments with other comments below (or above I don't remember). And to have a beautiful/consistent code style.

aabadie commented:
> The main thing I see is that you broke the 4-space indention in most files

@haukepetersen, I reverted those changes. Surprisingly, these blocks were already not at a multiple of 4 space position (29 sometimes). But this is not important.

aabadie commented:
@haukepetersen, is it ok for you here ?

aabadie commented:
it would be great if we could move forward on this one @haukepetersen

aabadie commented:
@haukepetersen, it looks you are not very responsive here. Is it fine if we dismiss your review ? I think I addressed all your comments and @smlng already ACK'ed this PR. Thanks

aabadie commented:
@haukepetersen, your comments are all addressed I think so do you have other comments here or is it ok if we dismiss you review ? I'd like to move forward with this one. Thanks !

kYc0o commented on boards/pba-d-01-kw2x/include/board.h:
```diff
@@ -60,80 +60,32 @@ extern "C"
-#define TMP006_ADDR         (0x41)
-/** @} */
+#define KW2XRF_PARAM_SPI           SPI_DEV(1)
+#define KW2XRF_PARAM_SPI_CLK       (SPI_CLK_10MHZ)
```
Just to be consistent you might want to remove the parenthesis on this define.

kYc0o replied:
Here and below btw.

aabadie replied:
Consistent with what ? I just tested a git grep git grep 'SPI_CLK_10MHZ' both syntaxes (with and without parenthesis) are used. I tend to prefer with parenthesis. Is there a consensus for that ?

kYc0o commented on drivers/include/isl29020.h:
```diff
@@ -70,20 +61,24 @@ typedef struct {
+    isl29020_params_t params;   /**< device initialization parameters */
+    float lux_fac;              /**< factor to calculate actual lux value */
+} isl29020_t;
+
```
Unrelated change?

kYc0o commented on drivers/include/l3g4200d.h:
```diff
@@ -76,45 +76,35 @@ typedef enum {
-} l3g4200d_t;
+    l3g4200d_mode_t mode;   /**< sampling frequency and bandwidth mode */
+    l3g4200d_scale_t scale; /**< scaling factor to normalize results */
+} l3g4200d_params_t;
```
I'm not sure these and the other struct changes belong to this PR. Can you explain them?

kYc0o commented on drivers/include/lps331ap.h:
```diff
@@ -62,18 +54,23 @@ typedef struct {
  * @return              -1 on error
  */
-int lps331ap_init(lps331ap_t *dev, i2c_t i2c, uint8_t address, lps331ap_rate_t rate);
+int lps331ap_init(lps331ap_t *dev, const lps331ap_params_t * params);
```
Here you have an API change, I'd say you should make another PR for these.

aabadie replied:
This is about unifying the param scheme in _all_ drivers, so I'd say it fits in the goal of this PR

kYc0o replied:
Yeah but here you're not modifying the params scheme but the API for this driver. I see it takes the params as input but what I argue here is that you modified the API, the params and structs just to fit with the new params scheme. Too much changes in a row...

aabadie replied:
> you modified the API, the params and structs just to fit with the new params scheme

Which again is the goal of this PR: unifying the way initialization parameters are defined **and** used.

> Too much changes in a row...

Maybe but required

aabadie commented:
> If this scheme is not present in some drivers, it should stay untouched, and in subsequent PRs being adapted.

On one hand, I agree with that. Initially I didn't want to flood GitHub with too many PRs and thought it was ok to that in a single PR.
On the other hand, it makes the API of all drivers consistent, which is what this PR is about. This way, by merging this PR, we won't stay in midstream

kYc0o commented on boards/mulle/board.c:
```diff
@@ -169,7 +169,7 @@ static int mulle_nvram_init(void)
 
     if (rec.u32 != MULLE_NVRAM_MAGIC_EXPECTED) {
-        int i;
+        uint16_t i;
```
This is unrelated too.

kYc0o commented:
> On one hand, I agree with that. Initially I didn't want to flood GitHub with too many PRs and thought it was ok to do that in a single PR.
On the other hand, it makes the API of all drivers consistent, which is what this PR is about. This way, by merging this PR, we won't stay in midstream

Still, it's too much lines (1000+) and as I pointed out previously some of the changes are almost unrelated to this PR. Thus I suggest to keep only the param modifications which don't modify the API and structs.

aabadie commented:
> Thus I suggest to keep only the param modifications which don't modify the API and structs.

OK

kYc0o commented on drivers/Makefile.dep:
```diff
@@ -347,3 +347,8 @@ ifneq (,$(filter tja1042,$(USEMODULE)))
+ifneq (,$(filter tsl2561,$(USEMODULE)))
+  FEATURES_REQUIRED += periph_i2c
+  USEMODULE += xtimer
+endif
```
Unrelated?

kYc0o commented on drivers/cc2420/include/cc2420_params.h:
```diff
@@ -30,54 +30,52 @@ extern "C" {
  */
 #ifndef CC2420_PARAM_SPI
-#define CC2420_PARAM_SPI        (SPI_DEV(0))
+#define CC2420_PARAM_SPI        SPI_DEV(0)
```
Here and below, you removed the parenthesis, thus I don't know what this is about, if keeping them, removing them... I don't know it there's a consensus on its use, but the changes in the PR suggest that you have a shared preference on them. I don't mind if they're there or not, it's just a bit strange to see this kind of changes.

kYc0o replied:
So, just to clarify, on which kind of defines do you prefer parenthesis ?

aabadie commented:
@kYc0o no need to spend to much on this PR, I'll close it after I open a bunch of new ones, containing the split changes you requested

kYc0o commented:
Ok thanks!

aabadie commented:
Closing in favor of all mentioned PRs above and the remaining ones that will come soon. Each one should be simpler to review @kYc0o ;)

aabadie commented:
> Closing in favor of all mentioned PRs above and the remaining ones that will come soon

I think I'm done but I'm not 100% sure. I admit I lost myself in all those branches and PRs...

> Each one should be simpler to review @kYc0o ;)

But now I'm wondering what was the simplest. It will be a lot of work reviewing all that, right ?
